zerver/actions/streams.py

```diff
@@ -353,14 +353,16 @@ def bulk_add_subs_to_db_with_logging(
     RealmAuditLog.objects.bulk_create(all_subscription_logs)
 
 
-def send_stream_creation_events_for_private_streams(
+def send_stream_creation_events_for_previously_invisible_streams(
     realm: Realm,
     stream_dict: Dict[int, Stream],
     altered_user_dict: Dict[int, Set[int]],
 ) -> None:
+    realm_non_guest_ids = set(active_non_guest_user_ids(realm.id))
```

Comment:
I have kept this here to avoid calling it in a loop, though it is cached. But this means this will be called even when there are no public streams.

Reply:
If it's cached in memcached, I'd put it in the loop.

Reply:
Even though this is cached, this is potentially a list of 10Ks of user IDs, and we're fetching it in a loop over all streams that are being added; that loop will make adding a new user really expensive. 

I think doing this query at all is the wrong approach -- we have in the callers where `altered_user_dict` is initialized the full user objects; we can just computed an `altered_guests` set at that time, and pass it through to this function, and then do `stream_user_ids & altered_guests` to compute the subset of the altered user IDs who are guest users.